### PR TITLE
TracePoint#enable w/o options traces attr reader/accessor/method calls

### PR DESCRIPTION
**Pull request by AndyObtiva**
Added specs for the following reported issues:
https://github.com/jruby/jruby/issues/6101
https://bugs.ruby-lang.org/issues/16383

### CODE REVIEW

**Comment by andrykonchin** [comment]
Looks like specs fail on Ruby 2.6 and 2.7

**Comment by AndyObtiva** [comment]
That is correct. One of the two new specs added should fail (the other passes). These were written to describe an issue reported in CRuby and JRuby as mentioned in the links. I think I was assuming this pull request would get merged after the issue is corrected.

**Comment by andrykonchin** [comment]
I see. Thank you for the clarification.

A spec which describes broken behavior can be declared as a bug with `ruby_bug` guard ([documentation](https://github.com/ruby/spec/blob/9d14d3fa3ef99ec6d922e8f87abc2cf18955ed50/CONTRIBUTING.md#guard-for-bug)). For instance:

```ruby
        ruby_bug "#14061", "2.4"..."2.5" do
          it "fires ensure block before returning while loads file" do
            File.write(@filename, <<-END_OF_CODE)
              ScratchPad << "before begin"
              begin
                return
              ensure
                ScratchPad << "within ensure"
              end
              ScratchPad << "after begin"
            END_OF_CODE

            load @filename
            ScratchPad.recorded.should == ["before begin", "within ensure"]
          end
        end
```

So we don't have to wait until the bug is fixed actually.

**Comment by eregon** [comment]
`ruby_bug` is a bit problematic here because it's not clear whether this is considered as a bug by MRI, and whether MRI will fix it soon or not.
But it's probably still OK to add the spec with a ruby_bug guard in the meantime.

File: core/tracepoint/enable_spec.rb [hunk 1]
```diff
@@ -97,11 +97,54 @@
   end
 
   ruby_version_is "2.6" do
-    describe 'target: option' do
-      before :each do
-        ScratchPad.record []
+    before :each do
```
Comment (eregon) [inline core/tracepoint/enable_spec.rb]: The new specs should not be under the `ruby_version_is` guard (behavior is same before/after 2.6) but outside of it, with their own `ruby_bug` guard.

**Comment by AndyObtiva** [comment]
The issue was acknowledged on the Ruby bug report and a fix has already been provided via a patch: 
https://bugs.ruby-lang.org/issues/16383

It's been also declared by matz as "good to have, but not a showstopper"

As suggested by @andrykonchin, I wrapped `ruby_bug` around the one failing spec, but after I followed @eregon's suggestion to move both new specs out of `ruby_version_is "2.6"`. 

All checks have passed after the latest commit. I look forward to any feedback.

Best regards

**Comment by eregon** [comment]
Yes, the issue is that patch causes a 30% slowdown on instance variable access in MRI, so I doubt they will merge it as-is, and it's an open question whether they will find another way with less overhead and when they would take the time for that.

@headius What do you think? Should we merge as-is? Or with a `ruby_version_is "3.0"` and hope it's fixed by then, otherwise keep increasing the version?
The problem is of course if MRI doesn't implement it, then other implementations have an unfair failure or have to implement it when MRI doesn't.

File: core/tracepoint/enable_spec.rb [hunk 2]
```diff
@@ -96,6 +96,55 @@
...
+      obj = klass.new
+
+      trace.enable do
+        obj.foo
+      end
+
+      ScratchPad.recorded.should == [:foo, :bar]
+    end
+
+    ruby_bug '#16383', '2.4'...'2.8' do
```
Comment (eregon) [inline core/tracepoint/enable_spec.rb]: Current MRI master reports as 2.8 but doesn't fix this bug, so this should be at least 3.0.

Comment (AndyObtiva) [reply core/tracepoint/enable_spec.rb]: OK, I changed 2.8 to 3.0

**Comment by AndyObtiva** [comment]
"Yes, the issue is that patch causes a 30% slowdown on instance variable access in MRI, so I doubt they will merge it as-is, and it's an open question whether they will find another way with less overhead and when they would take the time for that."

It's a bit early to judge that. They whipped up that solution quite quickly, so I assumed they'd take more time afterwards to come up with a better solution. I am quite confident of their problem solving skills, so I'm not worried.

File: core/tracepoint/enable_spec.rb [hunk 3]
```diff
@@ -96,7 +96,56 @@
...
+          obj.foo
+          obj.bar
+        end
+
+        ScratchPad.recorded.should == [:foo, :bar]
+      end
+    end
+  end
+
+    ruby_version_is "2.6" do
```
Comment (eregon) [inline core/tracepoint/enable_spec.rb]: This line is indented 2 spaces too deep

**Comment by eregon** [comment]
> I am quite confident of their problem solving skills, so I'm not worried.

That still doesn't give us a timeline unfortunately.
The main problem is that until MRI implements it, other implementations have an unfair failure or have to implement it when MRI doesn't.
Maybe that's not a very big problem but still that's been the reason `ruby_bug` has been avoided except when we knew it was already fixed in MRI master.

@headius Would JRuby implement it before MRI does for instance? You've generally been against `ruby_bug` where it isn't a clear and solved bug IIRC, so I would like your opinion.

In general, I think we should add specs only when the bug has been fixed/the feature has been implemented. Having it before is uncomfortable because it looks like a failure for other implementations, and we have to guess when it will be implemented in MRI which is obviously hard.

File: core/tracepoint/enable_spec.rb (hunk 3 above)
Comment (AndyObtiva) [reply core/tracepoint/enable_spec.rb]: Fixed.

**Comment by AndyObtiva** [comment]
Closing since CRuby's issue report doesn't seem to be moving quickly: https://bugs.ruby-lang.org/issues/16383

Closed the JRuby pull request too: https://github.com/jruby/jruby/issues/6101

**Pull request closed without merging**